Select a better set of coins

lontivero opened the pull request:
Closes #7452 Do not select more coins than strictly necessary.


Ping @MaxHillebrand

nopara73 commented on WalletWasabi/Blockchain/TransactionBuilding/SmartCoinSelector.cs:
```diff
@@ -65,21 +65,20 @@ public IEnumerable<ICoin> Select(IEnumerable<ICoin> coins, IMoney target)
+			.ToImmutableList();
 
-		var coinsToSpend = new List<SmartCoin>();
+		var coinsGroup = coinsInBestClusterByScript.Count < 10
```
Are you sure 10 is ok on all computers?

lontivero replied:
Sorry, I should have explained the change. Once we have the a set of coins, grouped by scripPubKey, in the the `coinsInBestClusterByScript` the question is, which of those coins should be selected? 

Before this PR the logic was to take coins until they were enough. After this PR the logic is find the best combination. The complex part in how to compute the combinations because we have the best possible solution that is find all the combinations that only works when there are only a few coins but then, what to do in case there are lots of coins in that cluster well I will prepare an example. 

> Are you sure 10 is ok on all computers?

No, I am not sure but that's the number we are using a few lines o code above so, it seems we assumed it was okay. Anyway, it would be good to play we this a bit more.

I am switching it back to draft until I have better answers and at least one example.

nopara73 reviewed (approved):
Concept ACK: instead of selecting the first set of good coin group we find, we select the coin group that has the fewest coins in it. However, the implementation is too complex for me to be 100% sure it's correct. I'll still ACK it because if it doesn't break our tests, I don't think we have anything to worry about.

lontivero commented:
When the number of coins in the best cluster is less than 10 the the best combination of coins is chosen from the combinatorics of up to 5 coins. In case the cluster contains more than 1 coins (lets say values: 1, 2, 3, ... 10) the coins are selected from the set: (1, 2, 3, .., 19, [1 + 2], [1 + 2 + 3], .... [1 + 2 + 3 +... + 10]).

molnard commented on WalletWasabi/Blockchain/TransactionBuilding/SmartCoinSelector.cs:
```diff
@@ -65,21 +65,21 @@ public IEnumerable<ICoin> Select(IEnumerable<ICoin> coins, IMoney target)
-		var coinsToSpend = new List<SmartCoin>();
+		var coinsGroup = coinsInBestClusterByScript.Count < 10
+			? coinsInBestClusterByScript.CombinationsWithoutRepetition(1, 5).Select(x => x.ToImmutableList())
+			: coinsInBestClusterByScript.Select(x => ImmutableList.Create(x))
```
I am not sure what is happening here. ImmutableList syntax worsens (for me) the readability. Can you add a few comments on what is happening here?

molnard reviewed (changes requested):
I am only able to do a superficial review, but it LGTM.

molnard reviewed (changes requested):
I think the PR is good. Can you modify the code to be more descriptive? I am not able to review it. Add some comments or introduce more local vars.

lontivero commented:
It's much simpler now. It looks like there are lots of changes but that's because I fixed the indentations and replaced anonymous types with tuples but in reality the only meaningful change is at the end of the Select method.

molnard reviewed (approved):
tACK
